include/xtensor/xfft.hpp

```diff
@@ -0,0 +1,317 @@
+                // Could probably be reimplemented using stride views...
+                for (std::size_t size = 2; size <= n; size *= 2)
+                {
+                    const std::size_t halfsize = size / 2;
+                    const std::size_t tablestep = n / size;
+                    for (std::size_t i = 0; i < n; i += size)
+                    {
+                        for (std::size_t j = i, k = 0; j < i + halfsize; j++, k += tablestep)
+                        {
+                            const auto temp = data(j + halfsize) * expTable(k);
+                            data(j + halfsize) = data(j) - temp;
+                            data(j) += temp;
+                        }
+                    }
+                    // Prevent overflow in 'size *= 2'
+                    if (size == n) {
+                        break;
+                    }
+                }
+                return data;
+            }
+
+            template<
+                typename T
+                >
+            auto transform_bluestein(xt::xarray<std::complex<T>> data)
+            {
+                // Find a power-of-2 convolution length m such that m >= n * 2 + 1
+                const std::size_t n = data.size();
+                std::size_t m = 1;
+                while (m / 2 <= n) 
+                {
+                    m *= 2;
+                }
+
+                // Trignometric table
+                auto expTable = xt::xtensor<std::complex<T>, 1>::from_shape({ n });
+                xt::xtensor<std::size_t, 1> i = xt::pow(xt::linspace<std::size_t>(0, n - 1, n), 2);
+                i %= (n * 2);
+                
+                //I would like to do this but there appears to be a compiler error in CPP14
+                //Works in CPP17 and MSVC CPP14
+                //Linker error when using numeric_constants in this header
+                //auto angles = xt::eval(::xt::numeric_constants<T>::PI * i / n);
+                auto angles = xt::eval(pi<T>() * i / n);
+                auto j = std::complex<T>(0, 1);
+                expTable = xt::exp(-angles * j);
+
+                // Temporary vectors and preprocessing
+                xt::xarray<std::complex<T>> av = xt::empty<std::complex<T>>({ m });
+                xt::view(av, xt::range(0, n)) = data * expTable;
+
+
+                xt::xarray<std::complex<T>> bv = xt::empty<std::complex<T>>({ m });
+                xt::view(bv, xt::range(0, n)) = xt::conj(expTable);
+                xt::view(bv, xt::range(-n + 1, xt::placeholders::_)) = xt::view(xt::conj(xt::flip(expTable)), xt::range(xt::placeholders::_, - 1));
+
+                // Convolution
+                auto cv = xt::fft::fft_convolve<T>(av, bv);
+
+                return xt::eval(xt::view(cv, xt::range(0, n)) * expTable);
+            }
+
+            template<
+                typename T
+            >
+            auto fft_impl(xt::xarray<std::complex<T>> data)
```

Comment:
Does this not need a reference or const &? Forgive me for noise if `xt::xarray` is (say) a smart pointer under the hood.

Reply:
This results in a deep copy of the contents using value semantics.

Reply:
Returning expressions can get you into trouble if the data they contain are captured by Ivalue reference then returned destructing the context and your reference to the underlying data... Dangling reference.

Reply:
It's a petty though that you currently loose static members if you are calling for example with `xt::xtensor`